File: examples/csharp/avalonia/NotesShowcase/Models/InMemoryNoteRepository.cs
using System.Text.Json;

namespace NotesShowcase.Models;

/// <summary>
/// Default <see cref="INoteRepository"/>: in-memory store with simulated I/O
/// delays (per plan §3.a.4). Thread-safe via a <see cref="SemaphoreSlim"/>
/// gate; safe to call from multiple async consumers.
/// </summary>
public sealed class InMemoryNoteRepository : INoteRepository
{
    private readonly List<NotebookModel> _notebooks;
    private readonly List<NoteModel> _notes;
    private readonly SemaphoreSlim _gate = new(1, 1);
    private readonly TimeSpan _loadAllDelay;
    private readonly TimeSpan _loadNotesDelay;
    private readonly TimeSpan _saveNoteDelay;
    private readonly TimeSpan _deleteNoteDelay;
    private readonly TimeSpan _addNotebookDelay;
    private readonly TimeSpan _exportDelay;

    /// <summary>
    /// Initializes the repository with <paramref name="seed"/>. Delay overrides
    /// are exposed for tests that need zero latency (default delays mirror the
    /// scenario doc §7: 300/150/200/120/120/150 ms).
    /// </summary>
    public InMemoryNoteRepository(
        (IReadOnlyList<NotebookModel> Notebooks, IReadOnlyList<NoteModel> Notes) seed,
        TimeSpan? loadAllDelay = null,
        TimeSpan? loadNotesDelay = null,
        TimeSpan? saveNoteDelay = null,
        TimeSpan? deleteNoteDelay = null,
        TimeSpan? addNotebookDelay = null,
        TimeSpan? exportDelay = null)
    {
        _notebooks = seed.Notebooks.ToList();
        _notes = seed.Notes.ToList();
        _loadAllDelay = loadAllDelay ?? TimeSpan.FromMilliseconds(300);
        _loadNotesDelay = loadNotesDelay ?? TimeSpan.FromMilliseconds(150);
        _saveNoteDelay = saveNoteDelay ?? TimeSpan.FromMilliseconds(200);
        _deleteNoteDelay = deleteNoteDelay ?? TimeSpan.FromMilliseconds(120);
        _addNotebookDelay = addNotebookDelay ?? TimeSpan.FromMilliseconds(120);
        _exportDelay = exportDelay ?? TimeSpan.FromMilliseconds(150);
    }

    /// <inheritdoc/>
    public async Task<(IReadOnlyList<NotebookModel> Notebooks, IReadOnlyList<NoteModel> Notes)> LoadAllAsync(
        CancellationToken ct = default)
    {
        await Task.Delay(_loadAllDelay, ct).ConfigureAwait(false);
        await _gate.WaitAsync(ct).ConfigureAwait(false);
        try
        {
            return (_notebooks.ToList(), _notes.ToList());
        }
        finally
        {
            _gate.Release();
        }
    }

    /// <inheritdoc/>
    public async Task<IReadOnlyList<NoteModel>> LoadNotesAsync(string notebookId, CancellationToken ct = default)
    {
        await Task.Delay(_loadNotesDelay, ct).ConfigureAwait(false);
        await _gate.WaitAsync(ct).ConfigureAwait(false);
        try
        {
            return _notes.Where(n => n.NotebookId == notebookId).ToList();
        }
        finally
        {
            _gate.Release();
        }
    }

    /// <inheritdoc/>
    public async Task SaveNoteAsync(NoteModel note, CancellationToken ct = default)
    {
        await Task.Delay(_saveNoteDelay, ct).ConfigureAwait(false);
        await _gate.WaitAsync(ct).ConfigureAwait(false);
        try
        {
            var idx = _notes.FindIndex(n => n.Id == note.Id);
            var stamped = note with { UpdatedAt = DateTimeOffset.UtcNow };
            if (idx >= 0) _notes[idx] = stamped;
            else _notes.Add(stamped);
        }
        finally
        {
            _gate.Release();
        }
    }

    /// <inheritdoc/>
    public async Task DeleteNoteAsync(string id, CancellationToken ct = default)
    {
        await Task.Delay(_deleteNoteDelay, ct).ConfigureAwait(false);
        await _gate.WaitAsync(ct).ConfigureAwait(false);
        try
        {
            _notes.RemoveAll(n => n.Id == id);
        }
        finally
        {
            _gate.Release();
        }
    }

    /// <inheritdoc/>
    public async Task AddNotebookAsync(NotebookModel notebook, CancellationToken ct = default)
    {
        await Task.Delay(_addNotebookDelay, ct).ConfigureAwait(false);
        await _gate.WaitAsync(ct).ConfigureAwait(false);
        try
        {
            _notebooks.Add(notebook);
        }
        finally
        {
            _gate.Release();
        }
    }

    /// <inheritdoc/>
    public async Task ExportAsync(
        IReadOnlyList<NotebookModel> notebooks,
        IReadOnlyList<NoteModel> notes,
        string path,
        CancellationToken ct = default)
    {
        await Task.Delay(_exportDelay, ct).ConfigureAwait(false);
        var payload = new { notebooks, notes };
        var json = JsonSerializer.Serialize(payload, new JsonSerializerOptions { WriteIndented = true });
        await File.WriteAllTextAsync(path, json, ct).ConfigureAwait(false);
    }
}

File: examples/swift/notes-showcase/Sources/NotesShowcaseCore/Repository/InMemoryNoteRepository.swift
import Foundation

/// Default `NoteRepository`: in-memory store with simulated I/O delays.
///
/// Concurrency-safe: Swift `actor` serialises all mutations, mirroring the
/// C# `SemaphoreSlim(1, 1)` gate. All methods are `async throws` per the
/// protocol; cancellation propagates via structured concurrency / `Task.sleep`.
///
/// Default delays mirror the scenario doc §7: 300 / 150 / 200 / 120 / 120 / 150 ms.
/// Pass zero overrides in tests that don't care about wall-clock timing.
public actor InMemoryNoteRepository: NoteRepository {

    // MARK: - State

    private var notebooks: [NotebookModel]
    private var notes: [NoteModel]

    // MARK: - Delay configuration

    private let loadAllDelay: TimeInterval
    private let loadNotesDelay: TimeInterval
    private let saveNoteDelay: TimeInterval
    private let deleteNoteDelay: TimeInterval
    private let addNotebookDelay: TimeInterval
    private let exportDelay: TimeInterval

    // MARK: - Injectable clock

    /// Returns the current timestamp. Injected for test determinism.
    private let now: () -> Date

    // MARK: - Init

    public init(
        seed: (notebooks: [NotebookModel], notes: [NoteModel]),
        loadAllDelay: TimeInterval = 0.3,
        loadNotesDelay: TimeInterval = 0.15,
        saveNoteDelay: TimeInterval = 0.2,
        deleteNoteDelay: TimeInterval = 0.12,
        addNotebookDelay: TimeInterval = 0.12,
        exportDelay: TimeInterval = 0.15,
        now: @escaping () -> Date = { Date() }
    ) {
        self.notebooks = seed.notebooks
        self.notes = seed.notes
        self.loadAllDelay = loadAllDelay
        self.loadNotesDelay = loadNotesDelay
        self.saveNoteDelay = saveNoteDelay
        self.deleteNoteDelay = deleteNoteDelay
        self.addNotebookDelay = addNotebookDelay
        self.exportDelay = exportDelay
        self.now = now
    }

    // MARK: - NoteRepository

    public func loadAll() async throws -> (notebooks: [NotebookModel], notes: [NoteModel]) {
        try await sleep(loadAllDelay)
        return (notebooks: notebooks, notes: notes)
    }

    public func loadNotes(notebookId: String) async throws -> [NoteModel] {
        try await sleep(loadNotesDelay)
        return notes.filter { $0.notebookId == notebookId }
    }

    public func saveNote(_ note: NoteModel) async throws {
        try await sleep(saveNoteDelay)
        let stamped = note.with(updatedAt: now())
        if let idx = notes.firstIndex(where: { $0.id == note.id }) {
            notes[idx] = stamped
        } else {
            notes.append(stamped)
        }
    }

    public func deleteNote(id: String) async throws {
        try await sleep(deleteNoteDelay)
        notes.removeAll { $0.id == id }
    }

    public func addNotebook(_ notebook: NotebookModel) async throws {
        try await sleep(addNotebookDelay)
        notebooks.append(notebook)
    }

    public func export(
        notebooks: [NotebookModel],
        notes: [NoteModel],
        path: String
    ) async throws {
        try await sleep(exportDelay)
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        encoder.dateEncodingStrategy = .iso8601
        let payload = ExportPayload(notebooks: notebooks, notes: notes)
        let data = try encoder.encode(payload)
        try data.write(to: URL(fileURLWithPath: path))
    }

    // MARK: - Private helpers

    private func sleep(_ interval: TimeInterval) async throws {
        guard interval > 0 else { return }
        try await Task.sleep(nanoseconds: UInt64(interval * 1_000_000_000))
    }
}

// MARK: - Export payload

private struct ExportPayload: Encodable {
    let notebooks: [NotebookModel]
    let notes: [NoteModel]
}
